File: config/config-annotation-processor/src/test/java/ru/tinkoff/kora/config/annotation/processor/ConfigExtractorGeneratorExtensionTest.java
package ru.tinkoff.kora.config.annotation.processor;


import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.annotation.processor.common.AbstractAnnotationProcessorTest;
import ru.tinkoff.kora.config.annotation.processor.processor.ConfigParserAnnotationProcessor;
import ru.tinkoff.kora.kora.app.annotation.processor.KoraAppProcessor;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class ConfigExtractorGeneratorExtensionTest extends AbstractAnnotationProcessorTest {
    @Test
    public void testExtensionAnnotatedRecord() throws Exception {
        compile(List.of(new KoraAppProcessor(), new ConfigParserAnnotationProcessor()), """
            @KoraApp
            public interface TestApp {
              @Root
              default String root(ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig> extractor) { return ""; }
            }
            """, """
            @ru.tinkoff.kora.config.common.annotation.ConfigValueExtractor
            public record TestConfig(String value){}
            """);
        compileResult.assertSuccess();

        var graph = loadGraph("TestApp");
        assertThat(graph.draw().getNodes()).hasSize(2);
    }

    @Test
    public void testExtensionAnnotatedInterface() throws Exception {
        compile(List.of(new KoraAppProcessor(), new ConfigParserAnnotationProcessor()), """
            @KoraApp
            public interface TestApp {
              @Root
              default String root(ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig> extractor) { return ""; }
            }
            """, """
            @ru.tinkoff.kora.config.common.annotation.ConfigValueExtractor
            public interface TestConfig {
              String value();
            }
            """);
        compileResult.assertSuccess();

        var graph = loadGraph("TestApp");
        assertThat(graph.draw().getNodes()).hasSize(2);
    }

    @Test
    public void testExtensionAnnotatedPojo() throws Exception {
        compile(List.of(new KoraAppProcessor(), new ConfigParserAnnotationProcessor()), """
            @KoraApp
            public interface TestApp {
              @Root
              default String root(ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig> extractor) { return ""; }
            }
            """, """
            @ru.tinkoff.kora.config.common.annotation.ConfigValueExtractor
            public class TestConfig {
              private String value;
              
              public String getValue() {
                return this.value;
              }
              
              public void setValue(String value) {
                this.value = value;
              }
              
              @Override
              public boolean equals(Object obj) {
                return obj instanceof TestConfig that && java.util.Objects.equals(this.value, that.value);
              }
              
              public int hashCode() { return java.util.Objects.hashCode(value); }
            }
            """);
        compileResult.assertSuccess();

        var graph = loadGraph("TestApp");
        assertThat(graph.draw().getNodes()).hasSize(2);
    }

    @Test
    public void testExtensionNonAnnotatedRecord() throws Exception {
        compile(List.of(new KoraAppProcessor(), new ConfigParserAnnotationProcessor()), """
            @KoraApp
            public interface TestApp {
              @Root
              default String root(ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig> extractor) { return ""; }
            }
            """, """
            public record TestConfig(String value){}
            """);
        compileResult.assertSuccess();

        var graph = loadGraph("TestApp");
        assertThat(graph.draw().getNodes()).hasSize(2);
    }

    @Test
    public void testExtensionNonAnnotatedInterface() throws Exception {
        compile(List.of(new KoraAppProcessor(), new ConfigParserAnnotationProcessor()), """
            @KoraApp
            public interface TestApp {
              @Root
              default String root(ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig> extractor) { return ""; }
            }
            """, """
            public interface TestConfig {
              String value();
            }
            """);
        compileResult.assertSuccess();

        var graph = loadGraph("TestApp");
        assertThat(graph.draw().getNodes()).hasSize(2);
    }

    @Test
    public void testExtensionNonAnnotatedPojo() throws Exception {
        compile(List.of(new KoraAppProcessor(), new ConfigParserAnnotationProcessor()), """
            @KoraApp
            public interface TestApp {
              @Root
              default String root(ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig> extractor) { return ""; }
            }
            """, """
            public class TestConfig {
              private String value;
              
              public String getValue() {
                return this.value;
              }
              
              public void setValue(String value) {
                this.value = value;
              }

              @Override
              public boolean equals(Object obj) {
                return obj instanceof TestConfig that && java.util.Objects.equals(this.value, that.value);
              }
              
              public int hashCode() { return java.util.Objects.hashCode(value); }
            }
            """);
        compileResult.assertSuccess();

        var graph = loadGraph("TestApp");
        assertThat(graph.draw().getNodes()).hasSize(2);
    }
}

File: config/config-symbol-processor/src/test/kotlin/ru/tinkoff/kora/config/symbol/processor/ConfigExtractorGeneratorExtensionTest.kt
package ru.tinkoff.kora.config.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.config.ksp.processor.ConfigParserSymbolProcessorProvider
import ru.tinkoff.kora.config.ksp.processor.ConfigSourceSymbolProcessorProvider
import ru.tinkoff.kora.kora.app.ksp.KoraAppProcessorProvider
import ru.tinkoff.kora.ksp.common.AbstractSymbolProcessorTest
import ru.tinkoff.kora.ksp.common.GraphUtil.toGraph

class ConfigExtractorGeneratorExtensionTest : AbstractSymbolProcessorTest() {
    @Test
    fun testExtensionAnnotatedDataClass() {
        compile(
            listOf(KoraAppProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), """
            @KoraApp
            interface TestApp {
              @Root
              fun root(extractor: ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig>) = ""
            }
            
            """.trimIndent(), """
            @ru.tinkoff.kora.config.common.annotation.ConfigValueExtractor
            data class TestConfig(val value: String)
            
            """.trimIndent()
        )

        compileResult.assertSuccess()
        val graph = loadClass("TestAppGraph").toGraph()
        assertThat(graph.draw.nodes)
            .hasSize(2)
    }

    @Test
    fun testExtensionAnnotatedInterface() {
        compile(
            listOf(KoraAppProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), """
            @KoraApp
            interface TestApp {
              @Root
              fun root(extractor: ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig>) = ""
            }
            
            """.trimIndent(), """
            @ru.tinkoff.kora.config.common.annotation.ConfigValueExtractor
            interface TestConfig {
              fun value(): String
            }
            
            """.trimIndent()
        )

        compileResult.assertSuccess()
        val graph = loadClass("TestAppGraph").toGraph()
        assertThat(graph.draw.nodes)
            .hasSize(2)
    }

    @Test
    fun testExtensionNonAnnotatedDataClass() {
        compile(
            listOf(KoraAppProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), """
            @KoraApp
            interface TestApp {
              @Root
              fun root(extractor: ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig>) = ""
            }
            """.trimIndent(), """
            data class TestConfig(val value: String){}
            
            """.trimIndent()
        )
        compileResult.assertSuccess()

        val graph = loadClass("TestAppGraph").toGraph()
        assertThat(graph.draw.nodes)
            .hasSize(2)
    }

    @Test
    fun testExtensionNonAnnotatedInterface() {
        compile(
            listOf(KoraAppProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), """
            @KoraApp
            interface TestApp {
              @Root
              fun root(extractor: ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<TestConfig>) = ""
            }
            
            """.trimIndent(), """
            interface TestConfig {
              fun value(): String
            }
            """.trimIndent()
        )
        compileResult.assertSuccess()

        val graph = loadClass("TestAppGraph").toGraph()
        assertThat(graph.draw.nodes)
            .hasSize(2)
    }

    @Test
    fun testExtensionNonAnnotatedRecord() {
        compile(
            listOf(KoraAppProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), """
            @KoraApp
            interface TestApp {
              @Root
              fun root(extractor: ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<ru.tinkoff.kora.config.symbol.processor.cases.RecordConfig>) = ""
            }
            
            """.trimIndent()
        )
        compileResult.assertSuccess()

        val graph = loadClass("TestAppGraph").toGraph()
        assertThat(graph.draw.nodes)
            .hasSize(2)
    }

    @Test
    fun testExtensionNonAnnotatedPojo() {
        compile(
            listOf(KoraAppProcessorProvider(), ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), """
            @KoraApp
            interface TestApp {
              @Root
              fun root(extractor: ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor<ru.tinkoff.kora.config.symbol.processor.cases.PojoConfig>) = ""
            }
            
            """.trimIndent()
        )
        compileResult.assertSuccess()

        val graph = loadClass("TestAppGraph").toGraph()
        assertThat(graph.draw.nodes)
            .hasSize(2)
    }
}
